src/store.c

```diff
@@ -231,6 +255,35 @@ static bool match_library_uri(struct pkcs11_module *pkcs11, struct parsed_uri *p
 	return true;
 }
 
+#define LOGIN_INFO_FMT	"PKCS#11 token \'%s\' in slot %lu (user pin)"
+static void prepare_login_info(struct store_ctx *sctx, struct dbg *dbg)
+{
+	size_t tlabel_size = 0;
+	char *tlabel = NULL;
+	CK_TOKEN_INFO ti;
+	size_t li_size;
+
+	if (pkcs11_get_token_info(&sctx->pctx->pkcs11, sctx->slot_id,
+				  &ti, dbg) == CKR_OK) {
+		tlabel_size = pkcs11_strlen(ti.label, sizeof(ti.label));
+		tlabel = OPENSSL_strndup((char *)ti.label, tlabel_size);
+	}
+
+	li_size = sizeof(LOGIN_INFO_FMT);
+	li_size += 20; /* max slot_id */
+	li_size += tlabel_size;
+
+	sctx->slot_login_info = OPENSSL_zalloc(li_size);
+	if (!sctx->slot_login_info) {
+		ps_dbg_warn(dbg, "continue without provider-specific login info");
+		return;
+	}
+
+	snprintf(sctx->slot_login_info, li_size, LOGIN_INFO_FMT,
```

Comment:
Instead of calculating the required space yourself, you could simply use asprintf() instead, it will allocate a large enough string for you for a given format string and values.